internal/services/pam/pam_test.go

```diff
@@ -815,40 +815,53 @@ func startSession(t *testing.T, client authd.PAMClient, username string) string
 	return sbResp.GetSessionId()
 }
 
-func TestMain(m *testing.M) {
-	if os.Getenv("GO_WANT_HELPER_PROCESS") != "" {
-		os.Exit(m.Run())
-	}
-
-	slog.SetLogLoggerLevel(slog.LevelDebug)
-
+// setupGlobalBrokerMock creates and points to a test-wide system bus, registering the mock broker on it.
+func setupGlobalBrokerMock() (cleanup func(), err error) {
 	// Start system bus mock.
 	busCleanup, err := testutils.StartSystemBusMock()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "%v\n", err)
-		os.Exit(1)
+		return cleanup, err
 	}
-	defer busCleanup()
+	cleanup = busCleanup
 
 	// Start brokers mock over dbus.
-	brokersConfPath, cleanup, err := initBrokers()
+	brokersConfPath, brokerCleanup, err := initBrokers()
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "%v\n", err)
-		os.Exit(1)
+		return cleanup, err
+	}
+
+	cleanup = func() {
+		brokerCleanup()
+		busCleanup()
 	}
-	defer cleanup()
 
 	// Get manager shared across grpc services.
 	globalBrokerManager, err = brokers.NewManager(context.Background(), brokersConfPath, nil)
 	if err != nil {
-		fmt.Fprintf(os.Stderr, "%v\n", err)
-		os.Exit(1)
+		return cleanup, err
 	}
 	mockBrokerGeneratedID, err = getMockBrokerGeneratedID(globalBrokerManager)
 	if err != nil {
+		return cleanup, err
+	}
+
+	return cleanup, nil
+}
+
+func TestMain(m *testing.M) {
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "" {
+		os.Exit(m.Run())
+	}
+
+	slog.SetLogLoggerLevel(slog.LevelDebug)
+
+	cleanup, err := setupGlobalBrokerMock()
+	if err != nil {
+		cleanup()
```

Comment:
This can be potentially nil if returning on very first error, no?

Reply:
you are correct, as I returned the cleanup func variable, this is nil by default and not the empty func